plugins/modem-manager/fu-mm-device.c

```diff
@@ -1605,6 +1653,76 @@ fu_mm_device_write_firmware_fdl(FuDevice *device, GBytes *fw, FuProgress *progre
 					      fw,
 					      error);
 }
+
+static gboolean
+fu_mm_device_write_firmware_dfota(FuDevice *device,
+				  GBytes *fw,
+				  FuProgress *progress,
+				  GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autofree gchar *upload_cmd = NULL;
+
+	locker = fu_device_locker_new_full(device,
+					   (FuDeviceLockerFunc)fu_mm_device_dfota_open,
+					   (FuDeviceLockerFunc)fu_mm_device_dfota_close,
+					   error);
+
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_mm_device_at_cmd(self, "AT+QFLST=?", TRUE, error)) {
+		g_prefix_error(error, "listing files not supported: ");
+		return FALSE;
+	}
+	/* if listing firmware file does not fail, there is an old firmware file to remove */
+	if (fu_mm_device_at_cmd(self,
+				"AT+QFLST=\"UFS:" FU_DFOTA_UPDATER_FILENAME "\"",
+				TRUE,
+				NULL)) {
+		g_debug("found orphaned firmware file - trying to delete it.");
+
+		if (!fu_mm_device_at_cmd(self,
+					 "AT+QFDEL=\"" FU_DFOTA_UPDATER_FILENAME "\"",
+					 TRUE,
+					 error)) {
+			g_prefix_error(error, "failed to delete existing firmware file: ");
+			return FALSE;
+		}
+	}
+	/* put the device into upload mode */
+	upload_cmd = g_strdup_printf("AT+QFUPL=\"%s\",%" G_GSIZE_FORMAT ",5,1",
+				     FU_DFOTA_UPDATER_FILENAME,
+				     g_bytes_get_size(fw));
+	if (!fu_mm_device_at_cmd(self, upload_cmd, TRUE, error)) {
+		g_prefix_error(error, "failed to enable upload mode: ");
+		return FALSE;
+	}
+
+	if (!fu_dfota_updater_upload_firmware(self->dfota_updater, fw, error))
+		return FALSE;
+
+	if (!fu_mm_device_at_cmd(self,
+				 "AT+QFOTADL=\"/data/ufs/" FU_DFOTA_UPDATER_FILENAME "\"",
+				 TRUE,
+				 error)) {
+		g_prefix_error(error, "failed to start update: ");
+		return FALSE;
+	}
+	/* wait and reopen port */
+	if (!fu_device_locker_close(locker, error))
+		return FALSE;
+	fu_device_sleep(device, FU_DFOTA_UPDATER_FOTA_RESTART_TIMEOUT_SECS * 1000);
+	locker = fu_device_locker_new_full(self,
+					   (FuDeviceLockerFunc)fu_mm_device_dfota_open,
+					   (FuDeviceLockerFunc)fu_mm_device_dfota_close,
+					   error);
+	if (locker == NULL)
+		return FALSE;
+
+	return fu_dfota_updater_write(self->dfota_updater, progress, device, fw, error);
+	/* uploaded firmware file is cleaned up automatically after device restart */
```

Comment:
at least _seems_ not always true with your above code.  Maybe it's worth cleaning up old file during probe or something?  You let me know what you think.
```c
	if (fu_mm_device_at_cmd(self,
				"AT+QFLST=\"UFS:" FU_DFOTA_UPDATER_FILENAME "\"",
				TRUE,
				NULL)) {

```

Reply:
If an existing firmware file has been found using the above `AT+QFLST` we delete it using `AT+QFDEL`.
Should the cleanup be moved to `fu_mm_device_probe()`?

Reply:
I've moved the old file cleanup into `fu_mm_device_setup` since `fu_mm_device_probe` does not seem ready for AT commands.